exp/controllers/gcpmachinepool_controller.go

```diff
@@ -0,0 +1,330 @@
+
+	// 	return ctrl.Result{}, nil
+	// }
+
+	// if !machinePoolScope.Cluster.Status.InfrastructureReady {
+	// 	log.Info("Cluster infrastructure is not ready yet")
+	// 	conditions.MarkFalse(machinePoolScope.GCPMachinePool, expinfrav1.MIGReadyCondition, infrav1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
+	// 	return ctrl.Result{}, nil
+	// }
+
+	// Make sure bootstrap data is available and populated
+	if machinePoolScope.MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName == nil {
+		log.Info("Bootstrap data secret reference is not yet available")
+		conditions.MarkFalse(machinePoolScope.GCPMachinePool, expinfrav1.MIGReadyCondition, infrav1.WaitingForBootstrapDataReason, clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, nil
+	}
+
+	instanceTemplateKey, err := instancetemplates.New(machinePoolScope).Reconcile(ctx)
+	if err != nil {
+		log.Error(err, "Error reconciling instanceTemplate")
+		// record.Warnf(machineScope.GCPMachine, "GCPMachineReconcile", "Reconcile error - %v", err)
+		conditions.MarkUnknown(machinePoolScope.GCPMachinePool, expinfrav1.InstanceTemplateReadyCondition, expinfrav1.InstanceTemplateNotFoundReason, "%s", err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// set the InstanceTemplateReadyCondition condition
+	conditions.MarkTrue(machinePoolScope.GCPMachinePool, expinfrav1.InstanceTemplateReadyCondition)
+
+	igm, err := instancegroupmanagers.New(machinePoolScope).Reconcile(ctx, instanceTemplateKey)
+	if err != nil {
+		log.Error(err, "Error reconciling instanceGroupManager")
+		// record.Warnf(machineScope.GCPMachine, "GCPMachineReconcile", "Reconcile error - %v", err)
+		conditions.MarkUnknown(machinePoolScope.GCPMachinePool, expinfrav1.MIGReadyCondition, expinfrav1.MIGNotFoundReason, "%s", err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// set the MIGReadyCondition condition
+	conditions.MarkTrue(machinePoolScope.GCPMachinePool, expinfrav1.MIGReadyCondition)
+
+	igmInstances, err := instancegroupmanagers.New(machinePoolScope).ListInstances(ctx, igm)
+	if err != nil {
+		log.Error(err, "Error listing instances in instanceGroupManager")
+		return ctrl.Result{}, err
+	}
+
+	providerIDList := make([]string, len(igmInstances))
+
+	for i, instance := range igmInstances {
+		var providerID string
+
+		// Convert instance URL to providerID format
+		u := instance.Instance
+		u = strings.TrimPrefix(u, "https://www.googleapis.com/compute/v1/")
+		tokens := strings.Split(u, "/")
+		if len(tokens) == 6 && tokens[0] == "projects" && tokens[2] == "zones" && tokens[4] == "instances" {
+			providerID = fmt.Sprintf("gce://%s/%s/%s", tokens[1], tokens[3], tokens[5])
+		} else {
+			return ctrl.Result{}, fmt.Errorf("unexpected instance URL format: %s", instance.Instance)
+		}
+
+		providerIDList[i] = providerID
+	}
+
+	// FUTURE: do we need to verify that the instances are actually running?
+	machinePoolScope.GCPMachinePool.Spec.ProviderIDList = providerIDList
```

Comment:
I really don't like this very much (it shouldn't be spec, it requires us to poll the cloud API), but it seems to be the MachinePool contract.